File: src/common/py/core/messaging/message.py
import abc
import collections.abc
import dataclasses
import typing
import uuid
from dataclasses import dataclass, field

from dataclasses_json import dataclass_json

from .channel import Channel
from .message_payload import MessagePayload
from ...utils import UnitID

MessageName = str
Trace = uuid.UUID


@dataclass_json
@dataclass(frozen=True, kw_only=True)
class Message(abc.ABC):
    """
    Base class for all messages.

    A message, besides its actual data, consists mainly of information from where it came and where it should go.

    This class also offers a useful decorator to easily declare new messages, like in the following example::

        @Message.define("msg/command")
        class MyCommand(Command):
            some_number: int = 0

    Attributes:
          name: The name of the message.
          origin: The initial source component of the message.
          sender: The component from where the message came from.
          target: Where the message should go to.
          hops: A list of components the message was sent through.
          trace: A unique trace identifying messages that logically belong together.
          api_key: An optional API key to access protected resources.
    """

    name: MessageName

    origin: UnitID
    sender: UnitID
    target: Channel

    hops: typing.List[UnitID] = field(default_factory=list)

    trace: Trace = field(default_factory=uuid.uuid4)

    api_key: str = ""

    @staticmethod
    def message_name() -> MessageName:
        """
        Retrieves the name of the message type on a message class basis.
        """
        return ""

    @staticmethod
    def is_protected() -> bool:
        """
        Whether this message is protected and thus requires an API key.
        """
        return False

    @staticmethod
    def define(name: str, is_protected: bool = False):
        """
        Defines a new message.

        The decorator takes care of wrapping the new class as a dataclass, passing the correct message
        name to its constructor. It also registers the new message type in the global ``MessageTypesCatalog``.

        Examples::

            @Message.define("msg/command")
            class MyCommand(Command):
                some_number: int = 0

        Args:
            name: The name of the message.
            is_protected: Whether the message requires an API key.
        """

        def decorator(cls):
            cls = dataclasses.dataclass(frozen=True, kw_only=True)(
                cls
            )  # Wrap the class in a dataclass
            __init__ = cls.__init__

            def __new_init__(self, *args, **kwargs):
                if "name" in kwargs:
                    del kwargs["name"]

                __init__(self, *args, name=MessageName(name), **kwargs)

            setattr(cls, "__init__", __new_init__)
            setattr(cls, "message_name", lambda *args, **kwargs: name)
            setattr(cls, "is_protected", lambda *args, **kwargs: is_protected)

            from .message_types_catalog import MessageTypesCatalog

            MessageTypesCatalog.register_item(name, cls)

            return cls

        return decorator

    def __post_init__(self):
        object.__setattr__(self, "_payload", MessagePayload())

    @property
    def payload(self) -> MessagePayload:
        """
        The message payload.
        """
        return typing.cast(MessagePayload, object.__getattribute__(self, "_payload"))

    def __str__(self) -> str:
        def cleanup(cur_obj, parent_obj, parent_key):
            if isinstance(cur_obj, str):
                from ...utils import human_readable_file_size

                if cur_obj.startswith("data:"):
                    parent_obj[parent_key] = (
                        cur_obj.split(",", 1)[0]
                        + f",<data:{human_readable_file_size(len(cur_obj))}>"
                    )
                elif len(cur_obj) > 256:
                    parent_obj[parent_key] = (
                        cur_obj[0:256]
                        + f"... [{human_readable_file_size(len(cur_obj))} total]"
                    )
            elif isinstance(cur_obj, dict):
                for key, value in cur_obj.items():
                    cleanup(value, cur_obj, key)
            elif isinstance(cur_obj, collections.abc.Sequence):
                for value in cur_obj:
                    cleanup(value, parent_obj, parent_key)

        obj_dict = self.to_dict()
        obj_dict["payload"] = str(self.payload)
        cleanup(obj_dict, None, None)
        return str(obj_dict)


# pylint: disable=invalid-name
MessageType = typing.TypeVar("MessageType", bound=Message)

File: src/common/web/core/messaging/Message.ts
import { plainToInstance, Type, Exclude } from "class-transformer";
// @ts-ignore
import { v4 as uuidv4 } from "uuid";
import { string } from "yup";

import { Channel } from "./Channel";
import { MessagePayload } from "./MessagePayload";
import { MessageTypesCatalog } from "./MessageTypesCatalog";
import { shortenDataStrings } from "../../utils/ObjectUtils";
import { type Constructable } from "../../utils/Types";
import { UnitID } from "../../utils/UnitID";

export type MessageCategory = string;
export type MessageName = string;
export type Trace = string;

/**
 * Base class for all messages.
 *
 * A message, besides its actual data, consists mainly of information from where it came and where it should go.
 *
 * This class also offers a useful decorator to easily declare new messages, like in the following example:
 * ```
 *     @Message.define("msg/command")
 *     class MyCommand extends Command {
 *          ...
 *     }
 * ```
 */
export abstract class Message {
    public static readonly Category: MessageCategory = "Message";

    public readonly name: string;
    // @ts-ignore
    @Type(() => UnitID)
    public readonly origin: UnitID;
    // @ts-ignore
    @Type(() => UnitID)
    public readonly sender: UnitID;
    // @ts-ignore
    @Type(() => Channel)
    public readonly target: Channel;

    // @ts-ignore
    @Type(() => UnitID)
    public readonly hops: UnitID[] = [];

    public readonly trace: Trace = "";

    public readonly api_key: string = "";

    @Exclude()
    public readonly payload: MessagePayload = new MessagePayload();

    /**
     * @param name - The name of the message.
     * @param origin - The initial source component of the message.
     * @param sender - The component from where the message came from.
     * @param target - Where the message should go to.
     * @param hops - A list of components the message was sent through.
     * @param trace - A unique trace identifying messages that logically belong together.
     * @param apiKey - An optional API key to access protected resources.
     */
    public constructor(name: string, origin: UnitID, sender: UnitID, target: Channel, hops: UnitID[] = [], trace: Trace = uuidv4(), apiKey: string = "") {
        this.name = name;
        this.origin = origin;
        this.sender = sender;
        this.target = target;
        this.hops = hops;
        this.trace = trace;
        this.api_key = apiKey;
    }

    /**
     * Converts this message to JSON.
     */
    public convertToJSON(): string {
        return JSON.stringify(this, (key: string, value: any): any | undefined => {
            if (key == "payload") {
                return undefined;
            }
            return value;
        });
    }

    /**
     * Creates a message from JSON data.
     *
     * @param msgType - The message type to construct.
     * @param data - The JSON data string.
     *
     * @returns - The created message.
     */
    public static convertFromJSON(msgType: ConstructableMessage, data: string): Message {
        let objData = JSON.parse(data);
        return plainToInstance(msgType, objData) as Message;
    }

    /**
     * Retrieves the name of the message type on a message class basis.
     */
    public static messageName(): MessageName {
        return "";
    }

    /**
     * The frontend never receives or issues protected messages.
     */
    public static isProtected(): boolean {
        return false;
    }

    /**
     * Defines a new message.
     *
     * The decorator takes care of wrapping the new class as a dataclass, passing the correct message
     * name to its constructor. It also registers the new message type in the global ``MessageTypesCatalog``.
     *
     * Examples:
     * ```
     *     @Message.define("msg/command")
     *     class MyCommand extends Command {
     *         ...
     *     }
     * ```
     *
     * @param name - The name of the message.
     */
    public static define(name: string): Function {
        return (ctor: Constructable): Constructable => {
            let newClass = class extends ctor {
                public constructor(...args: any[]) {
                    super(name, ...args);
                }

                public static messageName(): MessageName {
                    return name;
                }
            };

            MessageTypesCatalog.registerItem(name, newClass);

            return newClass;
        };
    }

    /**
     * Gets the global message category.
     */
    public abstract get messageCategory(): MessageCategory;

    /**
     * Gets the string representation of this message.
     */
    public toString(): string {
        // Remove all data fields from the output to not clutter the output
        const obj = JSON.parse(this.convertToJSON());
        obj["payload"] = this.payload.toString();
        return JSON.stringify(shortenDataStrings(obj));
    }
}

export interface ConstructableMessage<T extends Message = Message> {
    new (...args: any[]): T;

    messageName(): MessageName;
}
